pelican/__init__.py

```diff
@@ -162,46 +162,57 @@ def run(self):
                     pluralized_draft_pages,
                     time.time() - start_time))
 
-    def get_generator_classes(self):
-        generators = [ArticlesGenerator, PagesGenerator]
+    def _get_generator_classes(self):
+        discovered_generators = [
+            (ArticlesGenerator, "internal"),
+            (PagesGenerator, "internal")
+        ]
 
-        if self.settings['TEMPLATE_PAGES']:
-            generators.append(TemplatePagesGenerator)
-        if self.settings['OUTPUT_SOURCES']:
-            generators.append(SourceFileGenerator)
+        if self.settings["TEMPLATE_PAGES"]:
+            discovered_generators.append((TemplatePagesGenerator, "internal"))
 
-        for pair in signals.get_generators.send(self):
-            (funct, value) = pair
+        if self.settings["OUTPUT_SOURCES"]:
+            discovered_generators.append((SourceFileGenerator, "internal"))
 
-            if not isinstance(value, Iterable):
-                value = (value, )
+        for receiver, values in signals.get_generators.send(self):
+            if not isinstance(values, Iterable):
+                values = (values,)
 
-            for v in value:
-                if isinstance(v, type):
-                    logger.debug('Found generator: %s', v)
-                    generators.append(v)
+                discovered_generators.extend(
+                    [(generator, receiver.__module__) for generator in values]
+                )
```

Comment:
Is this meant to be inside of the `if not isinstance(...)` check? If `values` is already an iterable from `get_generators.send()` call then this code will never run (and they won't be added to `discovered_generators`).

Reply:
Oh, you are right. I missed the indent level in the diff 🤦 .

Reply:
That was my guess! 😢 I think the old behavior just needs to de-dent these three lines!

Reply:
Oh, snap, sorry guys, such a stupid mistake. I'll send a fix asap

Reply:
See #2823 Thanks @clokep for pointing it out